### PR TITLE
Issue 3703 - custom ticket - ticket image

### LINKED ISSUE #3703: Custom tickets - Image property
### Description
Using #3702 implement the image property 

### Goals
- [ ] I want to be able to be able to see a property of type image
- [ ] I want to be able to add a property of type image
- [ ] I want to be able to change a property of type image
- [ ] I want to be able to delete a property of type image

### Tasks
- [ ] Using #3702 reflect the image property value
- [ ] Create an edit image button with a dropdown to create a screenshot/upload image/delete image
- [ ] Save changes when creating a screenshot /uploading an image
- [ ] Delete the image when clicking in delete image

### Related Resources
[MockUp](https://www.figma.com/file/3FjNEOdQbcHwzquVZxUqMl/3D-Repo?node-id=788%3A5242)  (this is for viewpoint, but just ignore the edit viewpoint and go to viewpoint bits here)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #3703 

#### Description
New ticket allows for the image property to be part of the form and validates it correctly

#### Notes
There is a problem in the backend that is preventing the update of the image. The server responds with a 500 Error. This problem will be fixed as soon as the backend is fixed

### CODE REVIEW

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImageAction/actions/actions.component.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import { HiddenImageInput } from '@controls/formImage/hiddenImageUploader/hiddenImageUploader.styles';
+import { HiddenInputContainer } from '@controls/formImage/hiddenFileUploader/hiddenFileUploader.styles';
+import { ActionMenu, MenuItem, MenuItemDelete } from './actions.styles';
+import { TicketImageActionContext } from '../ticketImageActionContext';
+import { TicketImageAction } from '../ticketImageAction.component';
+
+export const GoToViewpointAction = () => (
+	<TicketImageAction disabled={false}>
+		<GoToViewpoinIcon />
+		<FormattedMessage id="viewer.card.ticket.viewpoint.action.goToViewpoint" defaultMessage="Go to viewpoint" />
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImageAction/actions/actions.component.tsx:37]: Should the id follow the same path as other messages in this file (i.e. ticketImage not ticket)

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImageAction/ticketImageActionContext.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { createContext } from 'react';
+
+interface TicketImageActionContextType {
+	imgSrc?: string;
+	setImgSrc?: (imgSrc) => void;
+	setImgFile?: (imgFile) => void;
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImageAction/ticketImageActionContext.tsx:22]: I'm not sure I understand the purpose behind having both setImgSrc and setImgFile

File: frontend/stories/inputs/file/HiddenFileUploader.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	argTypes: {
+		accept: {
+			type: 'string',
+		},
+	},
+	parameters: { controls: { exclude: ['className', 'onChange', 'id', 'disabled'] } },
+} as ComponentMeta<typeof HiddenFileUploader>;
+
+const ValidUploaderTemplate: ComponentStory<typeof HiddenFileUploader> = (args) => (
+	// eslint-disable-next-line
```
Comment (The-Daniel) [inline frontend/stories/inputs/file/HiddenFileUploader.stories.tsx:34]: I don't think this comment is needed any more

**Review by The-Daniel — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImageDisplayer/ticketImageDisplayer.component.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+					onChange={setImgSrc}
+				/>
+				<EnlargeContainer>
+					<EnlargeImageIcon />
+					<IconText>
+						<FormattedMessage id="viewer.cards.ticketImage.enlarge" defaultMessage="Enlarge" />
+					</IconText>
+				</EnlargeContainer>
+			</OverlappingContainer>
+			<Modal open={showLargePicture} onClickClose={() => setShowLargePicture(false)}>
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImageDisplayer/ticketImageDisplayer.component.tsx:48]: Image will not overflow the modal, removing the need for a (custom) scrollbar

File: frontend/src/v5/helpers/form.helper.ts [hunk 5]
```diff
@@ -35,7 +35,7 @@ export const dirtyValues = (
 };
 
 // eslint-disable-next-line max-len
-const isBasicValue = (value: any) => !!(value.toDate) || !_.isObject(value) || Array.isArray(value) || _.isString(value) || _.isDate(value);
+const isBasicValue = (value: any) => _.isNull(value) || !!(value.toDate) || !_.isObject(value) || Array.isArray(value) || _.isString(value) || _.isDate(value);
```
Comment (The-Daniel) [inline frontend/src/v5/helpers/form.helper.ts:38]: Is the isNull condition needed? Does `!_.isObject(value)` not cover this?

**Review by The-Daniel — APPROVED** [review]

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+type TicketImageProps = Omit<BasicTicketImageProps, 'onEmptyImageClick' | 'imgSrc' | 'children'> & {
+	value?: string;
+};
+export const TicketImage = ({ value, onChange, ...props }: TicketImageProps) => {
+	const [imgSrc, setImgSrc] = useState<string>();
+	const { props: { ticketId } } = useContext(CardContext);
+	const { teamspace, project, containerOrFederation } = useParams();
+	const isFederation = modelIsFederation(containerOrFederation);
+
+	const valueIsResourceId = () => {
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx:60]: this could be a general helper ``isResourceId(str)`` in a different file, I dont know if it should be for tickets or for the viewer in general.

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		const regexExp = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/gi;
+		return regexExp.test(value);
+	};
+
+	const getResourceUrl = () => {
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx:65]: this should be a more general function (or even hook if we need the federation and TeamSpace and project context)

`` getTicketResourceUrl(resource)``

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+			accept: getSupportedImageExtensions(),
+			onUpload: (file) => convertFileToImageSrc(file, handleImageChange),
+		});
+	};
+
+	useEffect(() => {
+		if (!imgSrc && valueIsResourceId()) {
+			setImgSrc(getResourceUrl());
+		}
+	}, [value]);
```
[the remark this answers is not in the document]
Comment (sanmont3drepo) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx:90]: Also line 73 wouldn't be needed because the onchange triggers a new value in react-hook-form

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/tickets/tickets.helpers.ts [hunk 9]
```diff
@@ -95,6 +95,7 @@ export const propertyValidator = ({ required, name, type }: PropertyDefinition)
 			validator = Yup.array();
 			break;
 		case 'oneOf':
+		case 'image':
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/tickets/tickets.helpers.ts:98]: The template allows the image to be marked as a required field, in that case the validation is required to prove that the field is not empty

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx (hunk 6 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx:60]: I will be moving it into `tickets.helper.ts` for now, as we will need it for other components outside the viewer, I will create a `viewer.helper.ts` move it there.

**Review by The-Daniel — APPROVED** [review]

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/basicTicketImage.styles.ts [hunk 10]
```diff
@@ -40,9 +30,19 @@ export const Container = styled.div`
 
 export const PropertyName = styled(Typography).attrs({
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/basicTicketImage.styles.ts]: I think we should use `<InputLabel />` (to stuck with materialuis family of components for inputs) here and in the pin details, changing the style to adjust to George's design.
Notice that `<InputLabel />` deals with the required prop.

File: frontend/src/v5/ui/themes/theme.ts [hunk 11]
```diff
@@ -982,6 +982,10 @@ export const theme = createTheme({
 						display: 'none',
 					},
 				},
+			},
+		},
+		MuiFormLabel: {
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/themes/theme.ts:987]: Why did you need to do this?

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/basicTicketImage.component.tsx [hunk 12]
```diff
@@ -15,11 +15,13 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 import { ProjectsHooksSelectors } from '@/v5/services/selectorsHooks/projectsSelectors.hooks';
+import { FormControl, FormHelperText } from '@mui/material';
 import { ActionsList, ActionsSide, Container, PropertyName } from './basicTicketImage.styles';
 import { TicketImageDisplayer } from './ticketImageDisplayer/ticketImageDisplayer.component';
 
 export type BasicTicketImageProps = {
 	imgSrc: string,
+	formError: any,
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/basicTicketImage.component.tsx]: would prefer if instead of formError, would have error and helperText to mimic materialui's interface. 

That way in the near future we can get rid of all the Form... files and just have one that would apply a ReactHookform adapter what knows how to pass from react-hook-form to materialuis inputs..

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/basicTicketImage.styles.ts [hunk 13]
```diff
@@ -15,34 +15,28 @@
...
+	${({ error, theme }) => error && css`
+		border: solid 1px ${theme.palette.error.main};
+		background-color: ${theme.palette.error.lightest};
+	`}
 `;
 
-export const PropertyName = styled(Typography).attrs({
-	variant: 'h5',
-})<{ required?: boolean }>`
+export const PropertyName = styled(FormLabel)`
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/basicTicketImage.styles.ts:36]: Not ken on the name of the component PropertyName. At these level these are not exactly properties, but just specific types of inputs.

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/themes/theme.ts (hunk 11 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/themes/theme.ts:987]: That is for the asterisk on the label when the input is required, so regardless whether we are using an `<InputLabel />` directly, or have it rendered via a `<Textfield />` (or other MUI inputs), the asterisk still gets the right styling

**Comment by sanmont3drepo** [comment]
There's a bug with this template
```
{
  "code": "IMS",
  "name": "Images"
  "config": {},
  "modules": [],
  "properties": [
    {
      "name": "An image",
      "type": "image"
    },
    {
      "name": "A required image",
      "type": "image",
      "required": true
    }
  ],
}
```

Repro:
- Create a ticket with two the two images already set
- Delete the first one (non required)
- Delete the second one (required)

What happens:
- You are able to delete it  (frontend error) 
-  Backend lets you delete it (backend error @ChristosTsiotsias )

What should happen
- Should trigger an invalid image input
- Backend you throw an error

atendedor

File: frontend/src/v5/validation/shared/validators.ts [hunk 14]
```diff
@@ -18,7 +18,7 @@ import { formatMessage } from '@/v5/services/intl';
 import { isNull } from 'lodash';
 import * as Yup from 'yup';
 
-export const trimmedString = Yup.string().transform((value) => value.trim());
+export const trimmedString = Yup.string().transform((value) => (value ?? '').trim());
```
Comment (sanmont3drepo) [inline frontend/src/v5/validation/shared/validators.ts:21]: I'm not sure this should be true. If Yup.string() can accept null/undefined . Then the tranfsorm bit here should do that too instead of converting it to string to then trim it.

It could bring unexpected consequences.

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+			<FormattedMessage id="viewer.card.ticketImage.action.editImage" defaultMessage="Edit image" />
+		</TicketImageAction>
+	);
+};
+
+type TicketImageProps = Omit<BasicTicketImageProps, 'onEmptyImageClick' | 'imgSrc' | 'children' | 'error' | 'helperText'> & {
+	value?: string,
+	formError?: any,
+};
+export const TicketImage = ({ value, onChange, formError, ...props }: TicketImageProps) => {
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/basicTicketImage/ticketImage/ticketImage.component.tsx:55]: TicketImage should ahve the Textfield interface.

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/imageProperty.component.tsx [hunk 16]
```diff
@@ -14,12 +14,35 @@
...
+export const ImageProperty = ({
+	property: { name: title, readOnly, required },
+	name,
+	defaultValue,
+	onBlur,
+	...props
+}: PropertyProps) => {
+	const { watch } = useFormContext();
+
+	useEffect(() => { onBlur?.(); }, [watch(name)]);
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/tickets/ticketsForm/properties/imageProperty.component.tsx:31]: Should be inside of TicketImage but not using react-hook-form references just 

``useffect(()=> onBlur() , [value])``

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by carmenfan — CHANGES_REQUESTED** [review]
Backend review, so this is for @ChristosTsiotsias

File: backend/src/v5/utils/helper/yup.js [hunk 17]
```diff
@@ -100,12 +100,13 @@ YupHelper.types.strings.name = Yup.string().min(1).max(35);
...
-	.test('Image', `must be of type ${fileUploads.imageExtensions.join(',')}`, async (buffer) => {
+YupHelper.types.embeddedImage = (isNullable) => Yup.mixed().transform((n, orgVal) => (orgVal ? Buffer.from(orgVal, 'base64') : n))
+	.test('Too large image', `Image must be smaller than ${fileUploads.resourceSizeLimit} Bytes`, (buffer) => !buffer || buffer.length <= fileUploads.resourceSizeLimit)
+	.test('Non supported type', `Image must be of type ${fileUploads.imageExtensions.join(',')}`, async (buffer) => {
 		if (!buffer) return true;
 		const ext = await fileExtensionFromBuffer(buffer);
 		return ext && fileUploads.imageExtensions.includes(ext.toLowerCase());
-	});
+	})
+	.test('Not null', 'Image cannot be null', (val) => isNullable || val !== null);
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/utils/helper/yup.js:110]: So because mixed() allows null I tried to change embeddedImage to object but Yup doesnt like it

image

**Review by sanmont3drepo — APPROVED** [review]
